clinic/src/clj/clinic/service/patient.clj

```diff
@@ -0,0 +1,113 @@
+(ns clinic.service.patient
+  (:require [clinic.fhir.client :as fc]
+            [clinic.fhir.utils :as fu]
+            [clojure.spec.alpha :as s]
+            [clojure.string :as string]
+            [java-time.api :as jt]))
+
+(def mrn-system "urn:nilenso:clinic:mrn")
+(def marital-status-system "http://hl7.org/fhir/ValueSet/marital-status")
+(def email-system "email")
+(def phone-system "phone")
+
+(def ^:private not-blank? (complement string/blank?))
+
+(defn- date? [v]
+  (try (jt/local-date v)
+       (catch Exception _ nil)))
+
+(s/def ::id (s/and string? not-blank?))
+(s/def ::mrn (s/and string? (partial re-matches #"\d{3}-\d{3}-\d{3}")))
+(s/def ::first-name (s/and string? not-blank?))
+(s/def ::last-name (s/and string? not-blank?))
+(s/def ::birth-date (s/and string? date?))
+(s/def ::gender #{"male" "female" "other" "unknown"})
+(s/def ::marital-status (s/nilable #{"A" "D" "I" "L" "M" "P" "S" "T" "U" "W" "UNK"}))
+(s/def ::email (s/nilable (s/and string? not-blank?)))
+(s/def ::phone (s/nilable (s/and string? not-blank?)))
+
+(s/def ::create-params
+  (s/keys :req-un [::first-name ::last-name ::birth-date ::gender]
+          :opt-un [::marital-status ::email ::phone]))
+
+(s/def ::patient
+  (s/keys :req-un [::id ::mrn ::first-name ::last-name ::birth-date ::gender]
+          :opt-un [::marital-status ::email ::phone]))
+
+(defn generate-mrn []
+  (String/format "%03d-%03d-%03d"
+                 (into-array [(rand-int 999)
+                              (rand-int 999)
+                              (rand-int 999)])))
+
+(defn- domain->fhir [params]
+  (let [resource (atom {:resourceType "Patient"
+                        :identifier [{:system mrn-system
+                                      :value (params :mrn)}]
+                        :name [{:family (params :last-name)
+                                :given [(params :first-name)]}]
+                        :birthDate (params :birth-date)
+                        :gender (params :gender)
+                        :telecom []
+                        :active true})]
+    (when (params :marital-status)
+      (swap! resource assoc :maritalStatus {:coding [{:system marital-status-system
+                                                      :code (params :marital-status)}]}))
+    (when (params :email)
+      (swap! resource update :telecom conj {:system email-system
+                                            :value (params :email)}))
+    (when (params :phone)
+      (swap! resource update :telecom conj {:system phone-system
+                                            :value (params :phone)}))
+    @resource))
```

Comment:
Don't think the atom is needed here.
Can just thread the resource through a series of assocs:
```
(-> resource-map
    (cond-> (:marital-status params)
        (assoc :maritalStatus {:coding ...})
     ...
     )
``` 
There are cases when this isn't feasible and you actually want to maintain state in an atom, but you don't need it here. Using an atom like this is analogous to assigning a value to a variable and iteratively/procedurally mutating it.

Reply:
Didn't know about `cond->`. Refactored it!